genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/UploadServiceImpl.java

```diff
@@ -0,0 +1,145 @@
+import com.netflix.genie.agent.execution.services.UploadService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.core.io.ResourceLoader;
+import org.springframework.core.io.WritableResource;
+import org.springframework.stereotype.Component;
+
+import javax.validation.constraints.NotNull;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+/**
+ * Implementation of UploadService.
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Component
+@Lazy
+@Slf4j
+class UploadServiceImpl implements UploadService {
+    private final ResourceLoader resourceLoader;
+
+    UploadServiceImpl(final ResourceLoader resourceLoader) {
+        this.resourceLoader = resourceLoader;
+    }
+
+    /**
+     * Compress the file and upload to a target location. For a directory
+     * all its contents are recursively compressed and uploaded to the target
+     * location
+     *
+     * @param file              file/directory to upload.
+     * @param targetLocationURI URI representing the target location for the file
+     * @throws UploadException if the upload fails
+     */
+
+    @Override
+    public void compressAndUpload(
+        @NotNull final File file,
+        @NotNull final URI targetLocationURI
+    ) throws UploadException {
+
+        final String uriString = targetLocationURI.toASCIIString();
+
+        log.debug("Uploading to target location: {}", uriString);
+
+        final WritableResource writableResource = (WritableResource) resourceLoader.getResource(uriString);
+
+        try (ZipOutputStream outputStream = new ZipOutputStream(writableResource.getOutputStream())) {
+            zipDirectory(file, file.getName(), outputStream);
+        } catch (Exception e) {
+            log.info("Error uploading to target location: {} ", uriString);
+            throw new UploadException("Error uploading file: " + file.getName(), e);
+        }
+    }
+
+    /**
+     * Zip a directory recursively.
+     *
+     * @param fileToZip         file/dir to zip
+     * @param nameInArchive     name for the file entry in archive
+     * @param zipOutputStream   destination output stream
+     * @throws IOException      in case of a problem
+     */
+    private void zipDirectory(
+        @NotNull final File fileToZip,
+        @NotNull final String nameInArchive,
+        @NotNull final ZipOutputStream zipOutputStream
+    ) throws IOException {
+
+        zipFile(fileToZip, nameInArchive, zipOutputStream);
+
+        if (fileToZip.isDirectory()) {
+            final File[] children = fileToZip.listFiles();
+            //Making findbugs happy. No need for this ideally as fileToZip is a directory and listFiles()
+            //is guaranteed not be null
+            if (children != null) {
+                for (File childFile : children) {
+                    zipDirectory(
```

Comment:
A bit confusing that you call `zipDirectory()` for regular files (non-directories).
Could find a better name for this helper.

Reply:
Makes sense. Will fix the name